### PR TITLE
Automatically load services in the src/ directory

### PR DESCRIPTION
**Pull request by aschempp**
This will finally add automatic autoloading for classes in the src folder. Due to Symfony's handling, it's a little bit complex. Here's how it works:

- Do nothing, if any existing `App\` service definition is available (e.g. in the `config/services.yml`)
- Otherwise, try to register the services directory. If a class is not found (a reflection class cannot be created, because the namespace is not registered in Composer), Symfony will throw an exception. Only the PHP loader implementation is able to catch these exceptions, that's why we use an `autoload.php` and not a YAML file.
- If services are successfully loaded, but the class does not exist in the file, the service definition will have an error (reflection exception, class found in autoloader but not in file).
- If all new services have errors, it means there is no class with `App\` namespace in the `src/` folder, but probably something else (e.g. `src/AppBundle/…`). Then remove all services and ignore all errors.
- If however only _some_ services have errors, we assume the user has made a mistake, e.g. misspelled the class name. We **do** want to show these errors then.

We might discuss whether we want to automatically tag controllers like Symfony does: https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/4.2/config/services.yaml#L22
I don't really see that point though, you can either extend `AbstractController` or tag the service yourself through annotations – something that's possible in Contao but not in Symfony (thanks to the ServiceAnnotationBundle 😉 )

This PR is complemented by https://github.com/contao/manager-plugin/pull/22 which makes sure the `App\` namespace is always registered in Composer. That's already the case for all regular managed edition because it was in the distribution `composer.json`, but it shouldn't be necessary to work.

### CODE REVIEW

**Comment by leofeyer** [comment]
This one is up for discussion, because I do not think that we should add any autoloading/autowiring magic besides the existing, well known and well documented Symfony defaults that you can enable in your `services.yml` file.

**Comment by fritzmg** [comment]
> This one is up for discussion, because I do not think that we should add any autoloading/autowiring magic besides the existing, well known and well documented Symfony defaults that you can enable in your `services.yml` file.

What do you think about adding a pre-configured `services.yml` to the manager bundle skeleton, like I originally suggested?

**Comment by leofeyer** [comment]
Where did you suggest this?

**Comment by fritzmg** [comment]
> Where did you suggest this?

Here: https://github.com/contao/managed-edition/issues/42

I accidentally opened it in the wrong repository back then.

**Comment by leofeyer** [comment]
I like it better. 🙈

**Comment by aschempp** [comment]
That won't work. If there is no `src` folder, or it contains anything else (e.g. an `AppBundle`), the container building will fail.

**Comment by aschempp** [comment]
But _implementing_ contao/managed-edition#42 is actually the point of this PR 😉

**Comment by fritzmg** [comment]
> That won't work. If there is no `src` folder, or it contains anything else (e.g. an `AppBundle`), the container building will fail.

True, but the skeleton could also contain a `src/.gitkeep` ... 😬. And if you create your own local bundle after running `composer create-project contao/managed-edition` you're likely to adjust the `services.yml` to your needs anyway.

That being said, I think I am more in favor of this PR now 😁 

> But _implementing_ [contao/managed-edition#42](https://github.com/contao/managed-edition/issues/42) is actually the point of this PR 😉

👍 



> We might discuss whether we want to automatically tag controllers like Symfony does: https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/4.2/config/services.yaml#L22
> I don't really see that point though, you can either extend `AbstractController` or tag the service yourself through annotations – something that's possible in Contao but not in Symfony (thanks to the ServiceAnnotationBundle 😉 )

Personally, in my projects, I use the ADR pattern and regular dependency injection via constructor and thus and only include the following:

```yml
App\Action\:
    resource: ../src/Action
    public: true
```

```php
namespace App\Action\ExampleAction;

use Doctrine\DBAL\Connection;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Annotation\Route;

/**
 * @Route("/example")
 */
class ExampleAction
{
    private $db;

    public function __construct(Connection $db)
    {
        $this->db = $db;
    }

    public function __invoke(Request $request): Response
    {
        return new Response('Hello World!');
    }
}
```

No tags needed. Service only needs to be public. May be you could add that? This way is also described in the getting started section of the documentation currently 😁

File: manager-bundle/src/HttpKernel/ContaoKernel.php [hunk 1]
```diff
@@ -216,6 +216,10 @@ public function registerContainerConfiguration(LoaderInterface $loader): void
         if ($servicesFile = $this->getConfigFile('services.yml')) {
             $loader->load($servicesFile);
         }
+
+        if (is_dir($this->getProjectDir().'/src')) {
+            $loader->load(__DIR__.'/../Resources/config/autoload.php');
```
Comment (leofeyer) [inline manager-bundle/src/HttpKernel/ContaoKernel.php]: As discussed in Mumble on January 16th, the file should be called `services.php` and go into the `src/Resources/skeleton/config/` folder.

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: manager-bundle/src/Resources/config/autoload.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    foreach ($originalDefinitions as $id => $definition) {
+        if (0 === strpos($id, 'App\\')) {
+            return;
+        }
+    }
+
+    $config = $configurator->services();
+
+    $config->defaults()->autowire(true);
+    $config->defaults()->autoconfigure(true);
```
Comment (leofeyer) [inline manager-bundle/src/Resources/config/autoload.php]: The `defaults()` method will create a new `DefaultsConfigurator` object upon each call. Maybe we can reuse the returned object?

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: manager-bundle/src/HttpKernel/ContaoKernel.php (hunk 1 above)
Comment (aschempp) [reply manager-bundle/src/HttpKernel/ContaoKernel.php]: Changed in 12cd852

File: manager-bundle/src/Resources/config/autoload.php (hunk 2 above)
Comment (aschempp) [reply manager-bundle/src/Resources/config/autoload.php]: Changed in b14a62f9d73fce75b7a33ca5869c80aeab68ebdc

**Review by leofeyer — APPROVED** [review]

File: manager-bundle/tests/Fixtures/HttpKernel/WithMixedNamespace/src/EventListener/InvalidListener.php [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Foo\EventListener;
+
+class InvalidListener
```
Comment (leofeyer) [inline manager-bundle/tests/Fixtures/HttpKernel/WithMixedNamespace/src/EventListener/InvalidListener.php:15]: This listener causes an issue that makes the `--prefer-lowest` tests fail: https://github.com/contao/contao/pull/1165/checks?check_run_id=393652779#step:6:40

File: manager-bundle/src/HttpKernel/ContaoKernel.php [hunk 4]
```diff
@@ -216,6 +216,10 @@ public function registerContainerConfiguration(LoaderInterface $loader): void
         if ($servicesFile = $this->getConfigFile('services.yml')) {
             $loader->load($servicesFile);
         }
+
+        if (is_dir($this->getProjectDir().'/src')) {
```
Comment (leofeyer) [inline manager-bundle/src/HttpKernel/ContaoKernel.php:220]: This condition is also true when unit testing the manager-bundle separately, therefore the bundles test fails: https://github.com/contao/contao/pull/1165/checks?check_run_id=393653023#step:4:888

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: manager-bundle/src/HttpKernel/ContaoKernel.php (hunk 4 above)
Comment (aschempp) [reply manager-bundle/src/HttpKernel/ContaoKernel.php:220]: When testing the bundle selaratey, its dependencies should still be installed? So core-bundle should be there as well as SecurityBundle ??

File: manager-bundle/tests/Fixtures/HttpKernel/WithMixedNamespace/src/EventListener/InvalidListener.php (hunk 3 above)
Comment (leofeyer) [reply manager-bundle/tests/Fixtures/HttpKernel/WithMixedNamespace/src/EventListener/InvalidListener.php:15]: Fixed in e669b8284c2f58c9f76ffdf9919ed8260bfaef9c.

**Review by leofeyer — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @aschempp.

**Pull request merged**